go/pkg/cosmos/txhistory.go

```diff
@@ -126,92 +162,416 @@ func Messages(msgs []sdk.Msg) []Message {
+func (c *txHistoryState) nextRecipTxHistoryPage() (*client.TxSearchResponse, error) {
+	c.cursor.TendermintRecvdPage++
+	res, err := c.tendermintHistoryRequest(histTxTypeRecipient)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to get receive transactions")
+		return nil, errors.Wrapf(err, "error getting tendermint recipient tx history")
 	}
 
-	totalReceive, err := strconv.ParseFloat(res.Result.TotalCount, 64)
+	return res, nil
+}
+
+func buildSentTxsQuery(address string) string {
+	queryTemplate := `"message.sender='%s'"`
+	return fmt.Sprintf(queryTemplate, address)
+}
+
+func buildRecvdTxsQuery(address string) string {
+	queryTemplate := `"transfer.recipient='%s'"`
+	return fmt.Sprintf(queryTemplate, address)
+}
+
+// given 2 sets of tx history, remove duplicates from the second set (recipient txs)
+func uniqueify(sender []client.TxSearchResponseResultTxs, recip []client.TxSearchResponseResultTxs) ([]client.TxSearchResponseResultTxs, []client.TxSearchResponseResultTxs) {
+	m := make(map[string]bool, len(sender)+len(recip))
+	filteredRecip := make([]client.TxSearchResponseResultTxs, 0, len(recip))
+	for i := 0; i < len(sender); i++ {
+		tx := sender[i]
+		m[*tx.Hash] = true
+	}
+	for i := 0; i < len(recip); i++ {
+		tx := recip[i]
+		if _, ok := m[*tx.Hash]; !ok {
+			m[*tx.Hash] = true
+			filteredRecip = append(filteredRecip, tx)
+		}
+	}
+	return sender, filteredRecip
+}
+
+func (c *Cursor) encode() *string {
+	if bytes, err := json.Marshal(c); err != nil {
+		logger.Errorf("error marshaling cursor from %#v: %s", c, err)
+		return nil
+	} else {
+		encoded := base64.StdEncoding.EncodeToString(bytes)
+		return &encoded
+	}
+}
+
+func (c *Cursor) decode(b64 string) error {
+	str, err := base64.StdEncoding.DecodeString(b64)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to parse totalReceive: %s", res.Result.TotalCount)
+		return errors.Wrapf(err, "error decoding base64 cursor %s", b64)
 	}
 
-	resTxs = append(resTxs, res.Result.Txs...)
+	if err := json.Unmarshal([]byte(str), c); err != nil {
+		return errors.Wrapf(err, "error marshaling cursor from %#v", c)
+	}
+	return nil
+}
 
-	// filter out duplicate transactions
-	seen := make(map[string]bool)
-	uniqueTxs := []client.TxSearchResponseResultTxs{}
-	for _, t := range resTxs {
-		if _, ok := seen[*t.Hash]; !ok {
-			uniqueTxs = append(uniqueTxs, t)
-			seen[*t.Hash] = true
+// read tendermint tx history
+func (s *txHistoryState) readTxHistory() (*TxHistory, error) {
+	// if we have cursor, restore state
+	if s.cursorSerialized != "" {
+		if err := s.cursor.decode(s.cursorSerialized); err != nil {
+			return nil, errors.Wrapf(err, "error decoding cursor %s", s.cursorSerialized)
+		} else {
+			// decrement by 1 so we start on last page we requested from tendermint
+			if s.cursor.TendermintSentPage > 0 {
+				s.cursor.TendermintSentPage--
```

Comment:
This seems weird to me and I would like to see whoever is calling or iterating the pages manage changing the page num. This is already a very complicated mutation state situation so making it as clear as possible who is changing the state and for what reason is going to go a long way to make things clear.

This does bring up the topic of even relying on such a heavily mutating state struct or if there would be a more functional way to accomplish this logic.

Reply:
This is the entry point, and should move to `GetTxHistory` as your prior comment suggested. Started dev in isolated function and left the skeleton/delegate.

It's initializing the cursor vars here, subtracting one to undo increment before prior page ended. might be able to change the way we increment to remove the need for that. Totally game for a cleaner/functional approach.